src/java.base/share/classes/java/lang/Thread.java

```diff
@@ -67,7 +66,13 @@
  * that execute {@link Runnable} tasks. {@linkplain  #start() Starting} a thread
  * schedules it to execute concurrently with the thread that caused it to start.
  * The newly started thread invokes the task's {@link Runnable#run() run} method.
- * Thread defines the {@link #join() join} method to wait for a thread to terminate.
+ *
+ * <p> A thread <i>terminates</i> if either its {@code run} method completes normally,
+ * or if its {@code run} method completes abruptly and the appropriate
+ * {@linkplain Thread.UncaughtExceptionHandler uncaught exception handler} completes
+ * normally or abruptly. With no code left to run, the thread has completed execution.
```

Comment:
I'm in two minds about introducing the UHE in paragraph 3 but probably okay as you kinda have to mention it when defining termination. There's a lot more to this story when you have agents in the picture but I'll not go there. I wonder if we can do something with the last sentence that includes the word "terminated". We use "terminated" in several places and it would be cleared if this sentence were to end with something like "... the thread has completes execution, and is terminated" (it could link to isAlive or Thread.State#TERMINATED).

Given that the join method is introduced as the method to wait for a thread to terminate then it could be part of this paragraph rather than a single sentence paragraph.

Reply:
I would also suggest that the join() sentence simply be the last sentence of the above paragraph:

> The join method can be be used to wait for a thread to terminate.

Reply:
The whole paragraph is about thread termination, so I don't think we need to say "terminate" again in the last sentence. We could mention isAlive or the TERMINATED state, but that brings in Thread.State which which require a discussion of the thread's life cycle.

The "join" sentence was kind of tacked on to the end of a paragraph in the earlier text; Alex suggested that I separate it into its own paragraph when we were working on the "termination" paragraph. Alex doesn't mind single-sentence paragraphs. Not my preference, but tacking the "join" sentence onto the end of the "termination" paragraph actually doesn't make any sense. Termination is something that happens to _this_ thread. The join method is called from _another_ thread, and this implicit shift of viewpoint would weaken the paragraph.

Frankly I think there needs to be a separate editorial pass over the Thread docs. I know they were just rewritten in JDK 19, but there are some odd things about it. The Thread.State life cycle isn't explained and isn't reference by methods such as start() and isAlive(). Also, various mentions of uncaught exception handling talk about a thread's abrupt termination. There isn't any such thing; a _method_ can complete normally or abruptly, and when a thread's run() method completes either way, the thread simply terminates.

Reply:
Thread.State was added (in Java 5) for monitoring and management purposes, it's not an API that most libraries or programs would use. It could be introduced in the class description and referenced from other APIs that change/test the state but it would encourage use beyond the original intent.

Reply:
What you have is okay although I would have preferred if the sentence on the join method was in the previous paragraph rather as it fits with termination.

Reply:
Plus one on the join() bit. You could also argue for mentioning isAlive() here too.

Reply:
Merged `join` sentence into previous paragraph per David's earlier suggestion. I didn't mention `isAlive` though, no need to complicate this further.